### PR TITLE
Added prometheus tests to prometheus validator

### PR DESCRIPTION
**Pull request by rporres**
Rule tests have been added in the same integration as the validations as
we needed the actual rule contents to run the tests, hence separating it
would have meant to fetch the rules twice. The integration has been
renamed to tester since test is more generic as it  includes validation.

Since rules can be templates, this means rules can have a different
shape for different namespaces. This force us to run the tests in every
namespace the rules appear. To make things simpler, I have limited the
number of rules per test to one as allowing for multiple rules
complicated significantly the run of the tests. If we see this is a
limitation, we can revisit this decision.

Related to https://gitlab.cee.redhat.com/service/app-interface/-/merge_requests/13709

Signed-off-by: Rafa Porres Molina <rporresm@redhat.com>

### CODE REVIEW

File: reconcile/prometheus_rules_tester.py [hunk 1]
```diff
@@ -0,0 +1,269 @@
...
+        rules[r['path']] = {}
+
+    for n in gqlapi.query(orb.NAMESPACES_QUERY)['namespaces']:
+        cluster = n['name']
+        namespace = n['cluster']['name']
+
+        if cluster_name and cluster != cluster_name:
+            continue
+
+        if not n['managedResourceTypes'] or \
```
Comment (maorfr) [inline reconcile/prometheus_rules_tester.py:68]: nit: use get?

File: reconcile/prometheus_rules_tester.py [hunk 2]
```diff
@@ -0,0 +1,269 @@
...
+                            f'{namespace} in cluster {cluster}')
+            continue
+
+        for r in openshift_resources:
+            path = r['path']
+            if path not in rules:
+                continue
+
+            # Or we will get an unexepected and confusing html_url annotation
+            if 'add_path_to_prom_rules' not in r:
```
Comment (maorfr) [inline reconcile/prometheus_rules_tester.py:84]: is `add_path_to_prom_rules` ever in `r`?

File: reconcile/prometheus_rules_tester.py [hunk 3]
```diff
@@ -0,0 +1,269 @@
...
+            msg = f"Error in test {f['path']}"
+            if 'cluster' in f:
+                msg += f" from namespace {f['namespace']} in cluster" \
+                       f"{f['cluster']}"
+            msg += f":  {f['check_result']}"
+
+            logging.warning(msg)
+
+    if invalid_rules or failed_tests:
+        sys.exit(1)
```
Comment (maorfr) [inline reconcile/prometheus_rules_tester.py:269]: please use reconcile.status.ExitCodes

File: reconcile/prometheus_rules_tester.py [hunk 4]
```diff
@@ -0,0 +1,269 @@
...
+
+def run(dry_run, thread_pool_size=10, cluster_name=None):
+    orb.QONTRACT_INTEGRATION = QONTRACT_INTEGRATION
+    orb.QONTRACT_INTEGRATION_VERSION = QONTRACT_INTEGRATION_VERSION
+
+    rules = get_prometheus_rules(cluster_name)
+    invalid_rules = check_prometheus_rules(rules, thread_pool_size)
+    if invalid_rules:
+        for i in invalid_rules:
+            logging.warning(f"Error in rule {i['path']} from namespace "
```
Comment (maorfr) [inline reconcile/prometheus_rules_tester.py:252]: if this eventually leads to exit with error, should we log error?

File: reconcile/prometheus_rules_tester.py [hunk 5]
```diff
@@ -0,0 +1,269 @@
...
+    failed_tests = check_prometheus_tests(tests, rules, thread_pool_size)
+    if failed_tests:
+        for f in failed_tests:
+            msg = f"Error in test {f['path']}"
+            if 'cluster' in f:
+                msg += f" from namespace {f['namespace']} in cluster" \
+                       f"{f['cluster']}"
+            msg += f":  {f['check_result']}"
+
+            logging.warning(msg)
```
Comment (maorfr) [inline reconcile/prometheus_rules_tester.py:266]: if this eventually leads to exit with error, should we log error?

**Review by maorfr — APPROVED** [review]
LGTM

i've added some minor comments, but you can merge as is as far as i'm concerned.
i love the fact that despite this being very complex, you were still able to write it with such simplicity. kudos for taking the time to go into all the corners cases.

File: reconcile/prometheus_rules_tester.py (hunk 2 above)
Comment (rporres) [reply reconcile/prometheus_rules_tester.py:84]: No, but it is set to `True` in `fetch_provider_resource`. That's why I needed to add this and make sure it is taken into account in `orb. fetch_openshift_resource`

File: reconcile/prometheus_rules_tester.py (hunk 4 above)
Comment (rporres) [reply reconcile/prometheus_rules_tester.py:252]: Yes, it makes sense.

File: reconcile/prometheus_rules_tester.py (hunk 5 above)
Comment (rporres) [reply reconcile/prometheus_rules_tester.py:266]: Yes, it makes sense.

File: reconcile/prometheus_rules_tester.py (hunk 3 above)
Comment (rporres) [reply reconcile/prometheus_rules_tester.py:269]: will do

File: reconcile/prometheus_rules_tester.py (hunk 1 above)
Comment (rporres) [reply reconcile/prometheus_rules_tester.py:68]: yes, either that or `not in` to avoid potential `KeyError` exceptions

Comment (rporres) [reply reconcile/prometheus_rules_tester.py]: fixed in a937393

File: reconcile/prometheus_rules_tester.py (hunk 5 above)
Comment (rporres) [reply reconcile/prometheus_rules_tester.py]: fixed in 248412f

File: reconcile/prometheus_rules_tester.py (hunk 4 above)
Comment (rporres) [reply reconcile/prometheus_rules_tester.py]: fixed in 248412f

File: reconcile/prometheus_rules_tester.py (hunk 3 above)
Comment (rporres) [reply reconcile/prometheus_rules_tester.py]: fixed in 1927ba1

**Pull request merged**